File: ext/lib/skkdic.rb
### skkserv/skkdic.rb --- rskkserv module for skkdic format dictionary.

## Copyright (C) 1997-2000  Shugo Maeda
## Copyright (C) 2000,2001  YAMASHITA Junji

## Author:	Shugo Maeda <shugo@aianet.ne.jp>
## Maintainer:	YAMASHITA Junji <ysjj@unixuser.org>
## Version:	2.3.2d

## This file is part of rskkserv.

## This program is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2, or (at
## your option) any later version.

## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.

### History:

## 2000/12/22:
# split from skkserv.rb.in

### Code:

require "skkserv/logger"

begin
  require "skkserv/skkdic.so"
rescue LoadError
  Logger::log(Logger::WARNING, "%s: %s", $0, $!)
#  raise RuntimeError, "can't load `skkserv/skkdic.so'" # for DEBUG
end

class SKKDic
  include FileTest

  OKURI_ARI_LABEL = ";; okuri-ari entries.\n"
  OKURI_NASI_LABEL = ";; okuri-nasi entries.\n"
  OKURI_ARI_REGEXP = /[\xa1-\xfe][\xa1-\xfe].*[a-z] $/n
  
  def initialize(path, cachedir, nocache)
    @path = path
    @file = open(@path)
    
    @cache_a = File.expand_path(File.basename(@path) + ".a", cachedir)
    @cache_n = File.expand_path(File.basename(@path) + ".n", cachedir)
    @nocache = nocache
    @okuri_ari_regexp = OKURI_ARI_REGEXP

    if !@nocache && new_cache?
      @okuri_ari_table = read_cache(@cache_a)
      @okuri_nasi_table = read_cache(@cache_n)
      if @okuri_ari_table.nil? || @okuri_nasi_table.nil?
	make_table
      end
    else
      make_table
    end

    if SKKDic.respond_to?(:search)
      @data = SKKDic.data(@path,
			  @okuri_ari_table.pack("i*"),
			  @okuri_nasi_table.pack("i*"))
      @okuri_ari_table = nil
      @okuri_nasi_table = nil
      GC.start

      @search = proc {|kana| SKKDic.search(kana, @data)}
    else
      @search = proc {|kana| search_rb(kana)}
    end
  end
  
  def new_cache?
    return (exist?(@cache_a) &&  File.ctime(@path) < File.ctime(@cache_a)) &&
           (exist?(@cache_n) &&  File.ctime(@path) < File.ctime(@cache_n))
  end

  def make_table
    Logger::log(Logger::INFO,
		"Making index table for dictionary \`%s\'...", @path)

    @okuri_ari_table = []
    @okuri_nasi_table = []
    for line in @file
      if line == OKURI_ARI_LABEL
	@okuri_ari_table.push(@file.pos)
	break
      end
    end
    for line in @file
      if line == OKURI_NASI_LABEL
	@okuri_nasi_table.push(@file.pos)
	break
      else
	@okuri_ari_table.push(@file.pos)
      end
    end
    @okuri_ari_table.pop
    @okuri_ari_table.reverse!
    for line in @file
      @okuri_nasi_table.push(@file.pos)
    end
    @okuri_nasi_table.pop
    
    Logger::log(Logger::INFO, "done.")

    unless @nocache
      write_cache(@okuri_ari_table, @cache_a)
      write_cache(@okuri_nasi_table, @cache_n)
    end
  end
  
  def write_cache(table, file)
    Logger::log(Logger::NOTICE, "Writing cache file \`%s\'...", file)

    f = nil
    begin
      f = open(file, "w")
      f.write(table.pack("i*"))
    rescue
      Logger::log(Logger::WARNING, "failed to write cache file \`%s\'.", file)
    ensure
      f.close if f
    end
    Logger::log(Logger::NOTICE, "done.")
  end
  
  def read_cache(file)
    Logger::log(Logger::INFO, "Reading cache file \`%s\'...", file)

    cache = nil
    open(file) do |f|
      cache = f.read.unpack("i*")
    end

    Logger::log(Logger::INFO, "done")

    return cache
  end
  
  def search(kana)
    @search.call(kana + " ")
  end

  private
  def search_rb(kana)
    if kana =~ @okuri_ari_regexp
      table = @okuri_ari_table
    else
      table = @okuri_nasi_table
    end
    
    left = 0
    right = table.length - 1
    while left <= right
      pos = (left + right) / 2

      @file.seek(table[pos], 0)
      buff = @file.readline
      case buff[0, kana.length] <=> kana
      when 0
	return buff[kana.length + 1 .. -3].split("/")
      when 1
	right = pos - 1
      when -1
	left = pos + 1
      end
    end
    
    return []
  end

  def to_s
    format('#<SKKDic: path="%s", cache=["%s", "%s"]>',
	   @path, @cache_a, @cache_n)
  end

  def self.create(path, options, config)
    SKKDic.new(path, config.skk_cache_dir, config.skk_no_cache)
  end
end

### skkserv/skkdic.rb ends here

File: ext/skkdic.c
/** ext/skkdic.c --- rskkserv module for skkdic.

 * Copyright (C) 2001-2005 YAMASHITA Junji

 * Author:	YAMASHITA Junji <ysjj@unixuser.org>
 * Version:	0.11

 * This file is part of rskkserv.

 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or (at
 * your option) any later version.

 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
**/

#define USE_MMAP 0
#define PROFILING 0

#if USE_MMAP
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#else
#include <stdio.h>
#endif /* USE_MMAP */

#if PROFILING
#include <sys/time.h>
#include <unistd.h>
#endif /* PROFILING */

#include <assert.h>
#include <string.h>
#include <ruby.h>

#if USE_MMAP
typedef char *JISYO;
#else
typedef FILE *JISYO;
#endif

struct skkdic_data {
  struct skkdic_tbl {
    int *ptr;
    int len;
  } a_tbl, n_tbl;
  int bufsize;
  JISYO jisyo;
#if USE_MMAP
  JISYO pos;
  off_t len;
  int fd;
#endif
};

#if USE_MMAP
#define skkdic_open(d,p) \
  do {\
    struct stat jisyo_stat;\
    (d)->fd = open(p, O_RDONLY);\
    fstat((d)->fd, &jisyo_stat);\
    (d)->len = jisyo_stat.st_size;\
    (d)->jisyo = (JISYO)mmap(0, (d)->len, PROT_READ, MAP_PRIVATE, (d)->fd, 0);\
  } while (0)

#define skkdic_close(d) \
  do {munmap((d)->jisyo, (d)->len); close((d)->fd);} while (0)

#define skkdic_seek(d,p) (d)->pos = (d)->jisyo + (p)
#define skkdic_getc(d) *((d)->pos++)

static inline int
skkdic_cmp(struct skkdic_data *skkdic, char *kana, int len)
{
  int result = strncmp(skkdic->pos, kana, len);
  skkdic->pos += len;
  return result;
}

#else /* ! USE_MMAP */
#define skkdic_open(d,p) (d)->jisyo = fopen(p, "r")
#define skkdic_close(d) fclose((d)->jisyo)
#define skkdic_seek(d,p) fseek((d)->jisyo, p, SEEK_SET)
#define skkdic_getc(d) fgetc((d)->jisyo)

static inline int
skkdic_cmp(struct skkdic_data *skkdic, char *kana, int len)
{
  unsigned char *p = (unsigned char*)kana;
  int c = EOF, n = 0, i = 0;
  while (len > i && (c=skkdic_getc(skkdic)) != EOF && (n=c-p[i++]) == 0)
    /* void */;
  if (c == EOF)
    return -1;
  return n;
}
#endif /* USE_MMAP */

static VALUE cSKKDic;

static VALUE fskkdic_s_data(VALUE, VALUE, VALUE, VALUE);
static VALUE fskkdic_s_search(VALUE, VALUE, VALUE);

#if PROFILING
static FILE *profile;
#endif

void Init_skkdic()
{
#if PROFILING
  profile = fopen("skkdic-mon.out", "w");
#endif

  rb_require("skkserv/skkdic.rb");

  cSKKDic = rb_define_class("SKKDic", rb_cObject);

  rb_define_singleton_method(cSKKDic, "data", fskkdic_s_data, 3);
  rb_define_singleton_method(cSKKDic, "search", fskkdic_s_search, 2);
}

#define FALSE 0
#define TRUE  1

static void fskkdic_data_free(struct skkdic_data *data)
{
  free(data->a_tbl.ptr);
  free(data->n_tbl.ptr);
  skkdic_close(data);
}

static VALUE
fskkdic_s_data(VALUE self, VALUE path, VALUE a_tbl, VALUE n_tbl)
{
  struct skkdic_data *data;
  VALUE retval;

  retval = Data_Make_Struct(rb_cData, struct skkdic_data, 0,
			    fskkdic_data_free, data);

  data->a_tbl.len = RSTRING_LEN(a_tbl) / sizeof(int);
  data->a_tbl.ptr = ALLOC_N(int, data->a_tbl.len);
  memcpy(data->a_tbl.ptr, RSTRING_PTR(a_tbl), RSTRING_LEN(a_tbl));

  data->n_tbl.len = RSTRING_LEN(n_tbl) / sizeof(int);
  data->n_tbl.ptr = ALLOC_N(int, data->n_tbl.len);
  memcpy(data->n_tbl.ptr, RSTRING_PTR(n_tbl), RSTRING_LEN(n_tbl));
  skkdic_open(data, StringValuePtr(path));
  data->bufsize = 16;

  return retval;
}

static int
is_okuri_ari(const char *kana, int len)
{
  const char c = kana[len-2];

  return (kana[0] & 0x80) && (c >= 'a' && c <= 'z');
}

static VALUE
fskkdic_s_search(VALUE self, VALUE vkana, VALUE vdata)
{
  char *kana = StringValuePtr(vkana);
  int kana_len = strlen(kana);
  int pos, left, right;
  int found;

  struct skkdic_data *skkdic;
  struct skkdic_tbl tbl;
  VALUE retval = rb_ary_new();

#if PROFILING
  struct timeval start_time, found_time, end_time;
  int loop_count = 0;
  gettimeofday(&start_time, NULL);
#endif

  Data_Get_Struct(vdata, struct skkdic_data, skkdic);

  tbl = is_okuri_ari(kana, kana_len) ? skkdic->a_tbl : skkdic->n_tbl;

  found = FALSE;
  left = 0;
  right = tbl.len - 1;

  while (left <= right) {
    int cmp;
#if PROFILING
    loop_count++;
#endif

    pos = (left + right) / 2;

    skkdic_seek(skkdic, tbl.ptr[pos]);
    cmp = skkdic_cmp(skkdic, kana, kana_len);
    if (cmp == 0) {
      found = TRUE;
      break;
    } else if (cmp > 0) {
      right = pos - 1;
    } else /* if (cmp < 0) */ {
      left = pos + 1;
    }
  }

#if PROFILING
  gettimeofday(&found_time, NULL);
#endif

  if (found) {
    int c = skkdic_getc(skkdic);

    assert(c == '/');
    for (c = skkdic_getc(skkdic); c != EOF && c != '\n'; c = skkdic_getc(skkdic)) {
      char *p = ALLOCA_N(char, skkdic->bufsize);
      int i = 0;
      p[i++] = c;
      while ((c=skkdic_getc(skkdic)) != '/') {
	if (i >= skkdic->bufsize) {
	  skkdic->bufsize *= 2;
	  p = memcpy(ALLOCA_N(char, skkdic->bufsize), p, i);
	}
	p[i++] = c;
      }
      rb_ary_push(retval, rb_str_new(p, i));
    }
  }

#if PROFILING
  {
    long search_usec, push_usec;

    gettimeofday(&end_time, NULL);

    search_usec = found_time.tv_usec - start_time.tv_usec;
    push_usec = end_time.tv_usec - found_time.tv_usec;
    if (search_usec < 0)
      search_usec += 1000 * 1000 * 1000;
    if (push_usec < 0)
      push_usec += 1000 * 1000 * 1000;

    fprintf(profile, "found: %3s, search usec (loop): %6d (%2d), ",
	    found ? "YES" : "NO", search_usec, loop_count);
    fprintf(profile, "build result: %3d\n", push_usec);
    fflush(profile);
  }
#endif

  return retval;
}

/** ext/skkdic.c ends here **/
